File: sdks/typescript/src/apiKeys.ts
import type { CallOptions, Transport } from "./transport.js";
import type { APIKey, CreatedAPIKey } from "./types.js";
import { requireArg } from "./validate.js";

interface APIKeyList {
  api_keys: APIKey[];
}

/**
 * ApiKeysService manages revocable API keys. These endpoints require the
 * master token; the server enforces this.
 */
export class ApiKeysService {
  constructor(private readonly t: Transport) {}

  /**
   * Issue a new API key. The raw secret is returned once in `key` and is
   * unrecoverable afterward.
   */
  async create(label?: string, opts: CallOptions = {}): Promise<CreatedAPIKey> {
    return this.t.json<CreatedAPIKey>("POST", "/v1/api-keys", {
      body: label ? { label } : {},
      signal: opts.signal,
    });
  }

  /** List API key metadata. */
  async list(opts: CallOptions = {}): Promise<APIKey[]> {
    const out = await this.t.json<APIKeyList>("GET", "/v1/api-keys", {
      signal: opts.signal,
    });
    return out.api_keys ?? [];
  }

  /** Revoke the API key with the given id. */
  async revoke(id: string, opts: CallOptions = {}): Promise<void> {
    requireArg(id, "id");
    await this.t.noContent("DELETE", `/v1/api-keys/${encodeURIComponent(id)}`, {
      signal: opts.signal,
    });
  }
}

File: sdks/rust/src/apikeys.rs
use std::sync::Arc;

use reqwest::Method;
use serde::{Deserialize, Serialize};

use crate::error::Error;
use crate::transport::{require, Transport};
use crate::types::{ApiKey, CreatedApiKey};

/// The API keys service, from [`Client::api_keys`](crate::Client::api_keys).
///
/// Every method requires the master token; the server enforces this.
#[derive(Clone)]
pub struct ApiKeys {
    t: Arc<Transport>,
}

impl ApiKeys {
    pub(crate) fn new(t: Arc<Transport>) -> Self {
        Self { t }
    }

    /// Issues a new API key. The raw secret is returned once in
    /// [`CreatedApiKey::key`] and is unrecoverable afterward.
    pub async fn create(&self, label: impl Into<String>) -> Result<CreatedApiKey, Error> {
        #[derive(Serialize)]
        struct Body {
            #[serde(skip_serializing_if = "String::is_empty")]
            label: String,
        }
        let request = self
            .t
            .request(Method::POST, &["v1", "api-keys"])
            .json(&Body {
                label: label.into(),
            });
        self.t.send_json(request).await
    }

    /// Returns the metadata for all API keys.
    pub async fn list(&self) -> Result<Vec<ApiKey>, Error> {
        #[derive(Deserialize)]
        struct Wire {
            #[serde(default)]
            api_keys: Vec<ApiKey>,
        }
        let request = self.t.request(Method::GET, &["v1", "api-keys"]);
        let wire: Wire = self.t.send_json(request).await?;
        Ok(wire.api_keys)
    }

    /// Deletes the API key with the given id.
    pub async fn revoke(&self, id: &str) -> Result<(), Error> {
        require(id, "id")?;
        let request = self.t.request(Method::DELETE, &["v1", "api-keys", id]);
        self.t.send_unit(request).await
    }
}
